lib/model/queue.ex

```diff
@@ -84,15 +89,41 @@ defmodule Queue do
 
   def new(queue_name, max_size) do
     OK.for do
-      primary_name <- check_queue(queue_name)
-      replica_name <- check_queue(Queue.replica_name(queue_name))
+      _ <- check_name(queue_name)
+      { primary_name, replica_name } <- complete_check(queue_name, &check_not_alive/1)
       primary_pid <- Horde.DynamicSupervisor.start_child(App.HordeSupervisor, {PrimaryQueue, [primary_name, max_size]})
       replica_pid <- Horde.DynamicSupervisor.start_child(App.HordeSupervisor, {ReplicaQueue, [replica_name, max_size]})
     after
       { primary_pid, replica_pid }
     end
   end
 
-  defp check_queue(queue_name),
+
+  def delete(queue_name) do
+    OK.for do
+      { primary_name, replica_name } <- complete_check(queue_name, &check_alive/1)
+      Horde.DynamicSupervisor.terminate_child(App.HordeSupervisor, whereis(primary_name))
+      Horde.DynamicSupervisor.terminate_child(App.HordeSupervisor, whereis(replica_name))
+    after
+      :deleted
+    end
+  end
+
+  defp check_name(queue_name),
+    do: OK.check({:ok, queue_name}, &(Queue.valid_name?(&1)), {:name_not_allowed, "Queue name #{inspect(queue_name)} is not allowed"})
+
+  defp check_not_alive(queue_name),
     do: OK.check({:ok, queue_name}, &(!Queue.alive?(&1)), {:queue_already_exists, "A queue named #{inspect(queue_name)} already exists"})
+
+  defp check_alive(queue_name),
+    do: OK.check({:ok, queue_name}, &(Queue.alive?(&1)), {:queue_not_found, "A queue named #{inspect(queue_name)} does not exist"})
+
+  defp complete_check(queue_name, check) do
+    OK.for do
```

Comment:
que es lo que hace el OK.for?

Reply:
Entiendo que como que secuencia las operaciones monadicas, es una de las formas de hacer una especie de "do notation". El nombre no me gusta la verdad. La otra forma de hacerlo es con `OK.try` pero esta ademas del `after` (que seria lo que retorna en caso exitoso) tiene otro parametro para catchear posibles errores. En este caso no es necesario.